File: sdk/js/cloudiy.mjs
/**
 * Cloudiy SDK for JavaScript (Node 18+ / browsers / edge runtimes).
 * Zero dependencies — plain fetch. Built for apps and AI agents.
 *
 *   import { CloudiyClient, PaymentRequiredError } from "./cloudiy.mjs";
 *
 *   const client = new CloudiyClient("127.0.0.1:8080", { token: "demo" });
 *   console.log(await client.info());
 *
 *   try {
 *     const r = await client.submit({ kernel: "vector_add", data: "1,2;3,4" });
 *     console.log(r.outputText);
 *   } catch (e) {
 *     if (e instanceof PaymentRequiredError) {
 *       // settle the USDC quote (Cloudiy escrow / x402), then retry:
 *       const r = await client.submit({
 *         kernel: "vector_add", data: "1,2;3,4", payment: e.demoPayment(),
 *       });
 *       console.log(r.outputText);
 *     } else throw e;
 *   }
 */

export class PaymentRequiredError extends Error {
  constructor(requirements) {
    const offer = requirements?.accepts?.[0] ?? {};
    const priceMicro = Number(offer.maxAmountRequired ?? 0);
    super(`payment required: ${priceMicro / 1e6} USDC to ${offer.payTo ?? "?"}`);
    this.name = "PaymentRequiredError";
    this.raw = requirements;
    this.priceMicroUsdc = priceMicro;
    this.priceUsdc = priceMicro / 1e6;
    this.payTo = offer.payTo ?? "";
    this.asset = offer.asset ?? "";
    this.network = offer.network ?? "";
    this.escrowProgram = offer.extra?.escrowProgram ?? "";
  }

  /** Base64 x402 payload for flow demos — real settlement uses the escrow. */
  demoPayment() {
    const payload = {
      x402Version: 1,
      scheme: "exact",
      network: this.network || "solana-devnet",
      payload: { note: "demo payment - settlement via Cloudiy escrow (devnet)" },
    };
    // UTF-8-safe base64 (btoa alone rejects non-latin1 characters)
    const bytes = new TextEncoder().encode(JSON.stringify(payload));
    let bin = "";
    for (const b of bytes) bin += String.fromCharCode(b);
    return btoa(bin);
  }
}

/** A transport/protocol failure talking to the node (unreachable node, HTTP
 * error, node-reported job error) — distinct from PaymentRequiredError (a quote)
 * and SignatureError (untrusted output). */
export class CloudiyError extends Error {
  constructor(message) {
    super(message);
    this.name = "CloudiyError";
  }
}

/** Thrown when a result's provider signature is missing, invalid, or not from
 * the expected node — the output must not be trusted. */
export class SignatureError extends CloudiyError {
  constructor(message) {
    super(message);
    this.name = "SignatureError";
  }
}

// --------------------------------------------------------------------------
// Result-signature verification (mirrors crates/common/src/sig.rs).
//
// The provider signs (job_id, sha256(input), sha256(output)) with its node key
// — the ed25519 key behind its iroh EndpointId (domain cloudiy/result/v2) — so
// a consumer can prove offline which node produced which output for which
// input. This SDK verifies that signature by default
// so an agent never silently trusts unverified output. Verification uses only
// public data, so a self-contained Ed25519 verify (BigInt point math + the
// runtime's SubtleCrypto for hashing) keeps the SDK zero-dependency across
// Node, browsers and edge runtimes.
// --------------------------------------------------------------------------

const _P = 2n ** 255n - 19n;
const _EDL = 2n ** 252n + 27742317777372353535851937790883648493n;
const _mod = (a) => ((a %= _P) < 0n ? a + _P : a);
const _pow = (b, e) => { b = _mod(b); let r = 1n; while (e > 0n) { if (e & 1n) r = _mod(r * b); b = _mod(b * b); e >>= 1n; } return r; };
const _inv = (a) => _pow(a, _P - 2n);
const _D = _mod(-121665n * _inv(121666n));
const _II = _pow(2n, (_P - 1n) / 4n);

function _recoverX(y, sign) {
  if (y >= _P) return null;
  let x2 = _mod((y * y - 1n) * _inv(_D * y * y + 1n));
  if (x2 === 0n) return sign ? null : 0n;
  let x = _pow(x2, (_P + 3n) / 8n);
  if (_mod(x * x - x2) !== 0n) x = _mod(x * _II);
  if (_mod(x * x - x2) !== 0n) return null;
  if ((x & 1n) !== BigInt(sign)) x = _P - x;
  return x;
}
const _GY = _mod(4n * _inv(5n));
const _GX = _recoverX(_GY, 0);
const _G = [_GX, _GY, 1n, _mod(_GX * _GY)];

function _ptAdd(P, Q) {
  const A = _mod((P[1] - P[0]) * (Q[1] - Q[0]));
  const B = _mod((P[1] + P[0]) * (Q[1] + Q[0]));
  const C = _mod(2n * P[3] * Q[3] * _D);
  const Dd = _mod(2n * P[2] * Q[2]);
  const E = B - A, F = Dd - C, G = Dd + C, H = B + A;
  return [_mod(E * F), _mod(G * H), _mod(F * G), _mod(E * H)];
}
function _ptMul(s, P) {
  let Q = [0n, 1n, 1n, 0n];
  while (s > 0n) { if (s & 1n) Q = _ptAdd(Q, P); P = _ptAdd(P, P); s >>= 1n; }
  return Q;
}
const _ptEq = (P, Q) => _mod(P[0] * Q[2] - Q[0] * P[2]) === 0n && _mod(P[1] * Q[2] - Q[1] * P[2]) === 0n;
function _toLE(bytes) { let n = 0n; for (let i = bytes.length - 1; i >= 0; i--) n = (n << 8n) | BigInt(bytes[i]); return n; }
function _ptDecompress(b) {
  if (b.length !== 32) return null;
  let y = _toLE(b); const sign = Number((y >> 255n) & 1n); y &= (1n << 255n) - 1n;
  const x = _recoverX(y, sign);
  return x === null ? null : [x, y, 1n, _mod(x * y)];
}
function _hex(s) {
  if (typeof s !== "string" || s.length % 2) return null;
  const out = new Uint8Array(s.length / 2);
  for (let i = 0; i < out.length; i++) { const b = parseInt(s.substr(i * 2, 2), 16); if (Number.isNaN(b)) return null; out[i] = b; }
  return out;
}
const _concat = (...arrs) => { const n = arrs.reduce((s, a) => s + a.length, 0); const o = new Uint8Array(n); let k = 0; for (const a of arrs) { o.set(a, k); k += a.length; } return o; };
const _sha = async (algo, bytes) => new Uint8Array(await crypto.subtle.digest(algo, bytes));

async function _ed25519Verify(pub, msg, sig) {
  if (pub.length !== 32 || sig.length !== 64) return false;
  const A = _ptDecompress(pub); if (!A) return false;
  const R = _ptDecompress(sig.slice(0, 32)); if (!R) return false;
  const S = _toLE(sig.slice(32)); if (S >= _EDL) return false;
  const h = _toLE(await _sha("SHA-512", _concat(sig.slice(0, 32), pub, msg))) % _EDL;
  return _ptEq(_ptMul(S, _G), _ptAdd(R, _ptMul(h, A)));
}

const _RESULT_DOMAIN = new TextEncoder().encode("cloudiy/result/v2");

/** True iff `signatureHex` is a valid provider signature over
 * (jobId, sha256(input), sha256(output)) by the node whose hex EndpointId is
 * `signedBy`. `input` must be the exact bytes submitted, so a provider that ran
 * a different prompt cannot produce a verifying signature. Same construction as
 * the Rust `cloudiy_common::sig` (v2). */
export async function verifyResult(signedBy, jobId, input, output, signatureHex) {
  const pub = _hex(signedBy), sig = _hex(signatureHex);
  if (!pub || !sig) return false;
  const msg = _concat(
    _RESULT_DOMAIN, new Uint8Array([0]),
    new TextEncoder().encode(jobId), new Uint8Array([0]),
    await _sha("SHA-256", input), new Uint8Array([0]),
    await _sha("SHA-256", output),
  );
  return _ed25519Verify(pub, msg, sig);
}

export class CloudiyClient {
  // `retries`: extra attempts idempotent GETs (info/health/status) make on a
  // transient failure (network error, timeout, HTTP 5xx). submit() is never
  // auto-retried — a paid job must not be resent and double-charged.
  constructor(node = "127.0.0.1:8080", { token, timeoutMs = 90_000, retries = 2 } = {}) {
    this.base = node.includes("://") ? node : `http://${node}`;
    this.token = token;
    this.timeoutMs = timeoutMs;
    this.retries = retries;
  }

  async #get(path) {
    const attempts = Math.max(1, this.retries + 1);
    let last;
    for (let i = 0; i < attempts; i++) {
      try {
        const res = await fetch(this.base + path, {
          signal: AbortSignal.timeout(this.timeoutMs),
        });
        // 5xx is transient (retry); 4xx is the caller's fault (don't).
        if (!res.ok) {
          if (res.status < 500 || i === attempts - 1) {
            throw new CloudiyError(`GET ${path} -> HTTP ${res.status}`);
          }
        } else {
          return res.json();
        }
      } catch (e) {
        if (e instanceof CloudiyError) throw e;
        // Network/abort error (fetch throws TypeError / AbortError).
        if (i === attempts - 1) {
          throw new CloudiyError(`cannot reach node at ${this.base} (${path}): ${e.message}`);
        }
        last = e;
      }
      await new Promise((r) => setTimeout(r, 200 * 2 ** i)); // 200ms, 400ms, …
    }
    throw last;
  }

  health() { return this.#get("/health"); }

  /** Node capabilities: GPU model, VRAM, price (USDC), escrow program. */
  info() { return this.#get("/info"); }

  status(jobId) { return this.#get(`/status/${jobId}`); }

  /**
   * Run a kernel on the node's GPU.
   * Throws PaymentRequiredError with the x402 quote when payment is needed.
   */
  async submit({ kernel, data, params = {}, token, payment, verify = true, expectPubkey = null } = {}) {
    const input = typeof data === "string" ? new TextEncoder().encode(data) : data;
    const headers = { "Content-Type": "application/json" };
    if (payment) headers["X-PAYMENT"] = payment;

    // A submit is not auto-retried (a paid job must not be resent), so a
    // connection failure is surfaced as a CloudiyError for the caller to handle.
    let res;
    try {
      res = await fetch(this.base + "/submit", {
        method: "POST",
        headers,
        signal: AbortSignal.timeout(this.timeoutMs),
        body: JSON.stringify({
          job_id: crypto.randomUUID(),
          kernel,
          input_data: Array.from(input),
          params,
          auth_token: token ?? this.token ?? "",
          consumer_pubkey: null,
          payment: payment ?? null,
        }),
      });
    } catch (e) {
      throw new CloudiyError(`cannot reach node at ${this.base} (/submit): ${e.message}`);
    }

    if (res.status === 402) throw new PaymentRequiredError(await res.json());
    if (!res.ok) throw new CloudiyError(`HTTP ${res.status}: ${await res.text()}`);

    const raw = await res.json();
    if (raw.status === "error") throw new CloudiyError(raw.error_message ?? "unknown error");

    let paymentReceipt = null;
    const receiptHeader = res.headers.get("x-payment-response");
    if (receiptHeader) {
      try { paymentReceipt = JSON.parse(atob(receiptHeader)); } catch { /* opaque */ }
    }

    const output = new Uint8Array(raw.output_data ?? []);
    const signature = raw.signature ?? null;
    const signedBy = raw.signed_by ?? null;

    // Verify the provider's signature over (job_id, sha256(input),
    // sha256(output)) by default — binds the output to THIS input.
    // With expectPubkey set, the signer must also BE that node (identity pin);
    // without it, a valid signature proves integrity — that `signedBy` signed
    // this exact output — but not that it is the node you intended.
    let signatureVerified = false;
    if (signature && signedBy && (expectPubkey === null || signedBy === expectPubkey)) {
      signatureVerified = await verifyResult(signedBy, raw.job_id, input, output, signature);
    }
    if (verify && !signatureVerified) {
      if (!signature || !signedBy) {
        throw new SignatureError("result was not signed by the provider — refusing to trust output (pass verify:false to accept unsigned)");
      }
      if (expectPubkey !== null && signedBy !== expectPubkey) {
        throw new SignatureError(`result signed by ${signedBy} but expected ${expectPubkey} — refusing to trust output`);
      }
      throw new SignatureError("invalid provider signature — result may be tampered; refusing to trust output");
    }

    return {
      jobId: raw.job_id,
      status: raw.status,
      output,
      outputText: new TextDecoder().decode(output),
      providerPubkey: raw.provider_pubkey ?? null,
      paymentReceipt,
      signature,
      signedBy,
      signatureVerified,
    };
  }
}

/** Function-calling tool schema so LLM agents can invoke Cloudiy GPU compute. */
export function asToolSchema(node = "127.0.0.1:8080") {
  return {
    name: "cloudiy_gpu_run",
    description:
      `Run a compute kernel on a decentralized GPU (Cloudiy network, node ${node}). ` +
      "Payment in USDC on Solana via x402. Kernels: vector_add ('a1,a2,...;b1,b2,...'), " +
      "matrix_mul ('m,k,n;A row-major;B row-major').",
    input_schema: {
      type: "object",
      properties: {
        kernel: { type: "string", enum: ["vector_add", "matrix_mul"] },
        data: { type: "string", description: "Kernel input in the documented format" },
      },
      required: ["kernel", "data"],
    },
  };
}

File: sdk/go/cloudiy.go
// Package cloudiy is a thin client for running GPU jobs on the Cloudiy network
// (USDC on Solana, x402). Zero third-party dependencies — stdlib only — built
// for apps and AI agents.
//
//	client := cloudiy.NewClient("127.0.0.1:8080")
//	info, _ := client.Info()
//	fmt.Println(info["price_usdc"])
//
//	res, err := client.Submit(cloudiy.SubmitOptions{Kernel: "vector_add", Data: []byte("1,2,3;4,5,6")})
//	var pr *cloudiy.PaymentRequiredError
//	if errors.As(err, &pr) { // x402: the node quoted its USDC price
//	    res, err = client.Submit(cloudiy.SubmitOptions{
//	        Kernel: "vector_add", Data: []byte("1,2,3;4,5,6"), Payment: pr.DemoPayment(),
//	    })
//	}
//	fmt.Println(string(res.Output), res.SignatureVerified)
//
// The provider's ed25519 result signature is verified by default: a missing or
// invalid signature returns a *SignatureError so an agent never trusts unverified
// output. Verification binds the output to the exact input submitted (domain
// cloudiy/result/v2), mirroring the Rust/Python/JS SDKs.
package cloudiy

import (
	"bytes"
	"crypto/ed25519"
	"crypto/rand"
	"crypto/sha256"
	"encoding/base64"
	"encoding/hex"
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"strings"
	"time"
)

// Version of the Go SDK.
const Version = "0.3.0"

const resultDomain = "cloudiy/result/v2"

// -- errors -----------------------------------------------------------------

// CloudiyError is a transport/protocol failure talking to the node (unreachable
// node, HTTP error, or a node-reported job error).
type CloudiyError struct{ Message string }

func (e *CloudiyError) Error() string { return e.Message }

// SignatureError means the result's provider signature was missing, invalid, or
// not from the expected node — the output must not be trusted.
type SignatureError struct{ Message string }

func (e *SignatureError) Error() string { return e.Message }

// PaymentRequiredError is the x402 quote returned when the node wants USDC
// before executing. Fields mirror the accepts[0] entry of the x402 requirements.
type PaymentRequiredError struct {
	Raw            map[string]any
	PriceMicroUSDC int64
	PriceUSDC      float64
	PayTo          string
	Asset          string
	Network        string
	EscrowProgram  string
}

func (e *PaymentRequiredError) Error() string {
	return fmt.Sprintf("payment required: %g USDC to %s (escrow %s)",
		e.PriceUSDC, e.PayTo, e.EscrowProgram)
}

// DemoPayment returns a base64 x402 payload for flow demos — real settlement
// uses the Cloudiy escrow program on Solana devnet.
func (e *PaymentRequiredError) DemoPayment() string {
	network := e.Network
	if network == "" {
		network = "solana-devnet"
	}
	payload, _ := json.Marshal(map[string]any{
		"x402Version": 1,
		"scheme":      "exact",
		"network":     network,
		"payload":     map[string]any{"note": "demo payment — settlement via Cloudiy escrow (devnet)"},
	})
	return base64.StdEncoding.EncodeToString(payload)
}

func newPaymentRequired(raw map[string]any) *PaymentRequiredError {
	var offer map[string]any
	if accepts, ok := raw["accepts"].([]any); ok && len(accepts) > 0 {
		offer, _ = accepts[0].(map[string]any)
	}
	if offer == nil {
		offer = map[string]any{}
	}
	priceMicro := int64(asFloat(offer["maxAmountRequired"]))
	escrow := ""
	if extra, ok := offer["extra"].(map[string]any); ok {
		escrow, _ = extra["escrowProgram"].(string)
	}
	return &PaymentRequiredError{
		Raw:            raw,
		PriceMicroUSDC: priceMicro,
		PriceUSDC:      float64(priceMicro) / 1_000_000,
		PayTo:          asString(offer["payTo"]),
		Asset:          asString(offer["asset"]),
		Network:        asString(offer["network"]),
		EscrowProgram:  escrow,
	}
}

// -- signature verification -------------------------------------------------

// resultSigningPayload builds the exact bytes the provider signs:
// domain ‖ 0 ‖ job_id ‖ 0 ‖ sha256(input) ‖ 0 ‖ sha256(output).
func resultSigningPayload(jobID string, input, output []byte) []byte {
	inH := sha256.Sum256(input)
	outH := sha256.Sum256(output)
	var b bytes.Buffer
	b.WriteString(resultDomain)
	b.WriteByte(0)
	b.WriteString(jobID)
	b.WriteByte(0)
	b.Write(inH[:])
	b.WriteByte(0)
	b.Write(outH[:])
	return b.Bytes()
}

// VerifyResult reports whether signatureHex is a valid provider signature over
// (jobID, sha256(input), sha256(output)) by the node whose hex EndpointId is
// signedBy. input must be the exact bytes submitted, so a provider that ran a
// different prompt cannot produce a verifying signature. Same construction as
// the Rust cloudiy_common::sig (v2). Uses the stdlib crypto/ed25519.
func VerifyResult(signedBy, jobID string, input, output []byte, signatureHex string) bool {
	pub, err := hex.DecodeString(signedBy)
	if err != nil || len(pub) != ed25519.PublicKeySize {
		return false
	}
	sig, err := hex.DecodeString(signatureHex)
	if err != nil || len(sig) != ed25519.SignatureSize {
		return false
	}
	return ed25519.Verify(pub, resultSigningPayload(jobID, input, output), sig)
}

// -- client -----------------------------------------------------------------

// Client is an HTTP client for a Cloudiy node (`cloudiy share` exposes the API).
// For the P2P transport (dial-by-NodeID, NAT traversal) use the Rust SDK; this
// client targets the node's HTTP endpoint — ideal for agents and backends.
type Client struct {
	base  string
	Token string
	// Retries is the number of extra attempts idempotent GETs (Info/Health/
	// Status) make on a transient failure (connection error, timeout, HTTP 5xx).
	// Submit is never auto-retried — a paid job must not be resent and
	// double-charged.
	Retries int
	http    *http.Client
}

// NewClient targets a node ("host:port" or a full URL). Timeout defaults to 90s
// (GPU jobs are bounded to 60s node-side); override via Client.SetTimeout.
func NewClient(node string) *Client {
	base := node
	if !strings.Contains(node, "://") {
		base = "http://" + node
	}
	return &Client{
		base:    base,
		Retries: 2,
		http:    &http.Client{Timeout: 90 * time.Second},
	}
}

// SetTimeout sets the per-request timeout.
func (c *Client) SetTimeout(d time.Duration) { c.http.Timeout = d }

// WithToken sets the auth token / access code and returns the client for chaining.
func (c *Client) WithToken(token string) *Client { c.Token = token; return c }

// get fetches path and decodes JSON, retrying transient failures (connection
// error, timeout, HTTP 5xx) with exponential backoff. Idempotent, so safe.
func (c *Client) get(path string) (map[string]any, error) {
	attempts := c.Retries + 1
	if attempts < 1 {
		attempts = 1
	}
	var lastErr error
	for i := 0; i < attempts; i++ {
		res, err := c.http.Get(c.base + path)
		if err != nil {
			lastErr = &CloudiyError{fmt.Sprintf("cannot reach node at %s (%s): %v", c.base, path, err)}
		} else {
			body, _ := io.ReadAll(res.Body)
			res.Body.Close()
			if res.StatusCode < 300 {
				var out map[string]any
				if err := json.Unmarshal(body, &out); err != nil {
					return nil, &CloudiyError{fmt.Sprintf("GET %s: bad JSON: %v", path, err)}
				}
				return out, nil
			}
			// 5xx is transient (retry); 4xx is the caller's fault (don't).
			if res.StatusCode < 500 {
				return nil, &CloudiyError{fmt.Sprintf("GET %s -> HTTP %d", path, res.StatusCode)}
			}
			lastErr = &CloudiyError{fmt.Sprintf("GET %s -> HTTP %d", path, res.StatusCode)}
		}
		if i < attempts-1 {
			time.Sleep(time.Duration(200*(1<<i)) * time.Millisecond) // 200ms, 400ms, …
		}
	}
	return nil, lastErr
}

// Health returns the node's liveness/uptime.
func (c *Client) Health() (map[string]any, error) { return c.get("/health") }

// Info returns node capabilities: GPU model, VRAM, price (USDC), escrow program.
func (c *Client) Info() (map[string]any, error) { return c.get("/info") }

// Status returns the status of a job by id.
func (c *Client) Status(jobID string) (map[string]any, error) { return c.get("/status/" + jobID) }

// JobResult is a completed job's output and its verification state.
type JobResult struct {
	JobID    string
	Output   []byte
	Status   string
	ProviderPubkey string
	PaymentReceipt map[string]any
	// SignatureVerified is true when the result carried a valid ed25519
	// signature from the provider node (and matched ExpectPubkey when pinned).
	SignatureVerified bool
	// Signature is the hex ed25519 result signature.
	Signature string
	// SignedBy is the hex node key (iroh EndpointId) that produced the signature.
	SignedBy string
}

// OutputText returns the output decoded as UTF-8 text.
func (r *JobResult) OutputText() string { return string(r.Output) }

// SubmitOptions configures a Submit call.
type SubmitOptions struct {
	Kernel string
	Data   []byte
	Params map[string]string
	// Token overrides the client's token for this call.
	Token string
	// Payment is a base64 x402 payment payload.
	Payment string
	// Verify verifies the provider's result signature. Defaults to true; set
	// NoVerify to accept unsigned results from a trusted-local/demo node.
	NoVerify bool
	// ExpectPubkey pins the provider's hex node identity.
	ExpectPubkey string
}

// Submit runs a kernel on the node's GPU. It returns a *PaymentRequiredError
// (via errors.As) with the x402 quote when the node wants USDC and no valid
// Payment/Token was given. The provider's ed25519 result signature is verified
// by default: a missing or invalid signature returns a *SignatureError. Submit
// is not auto-retried (a paid job must not be resent and double-charged).
func (c *Client) Submit(opts SubmitOptions) (*JobResult, error) {
	params := opts.Params
	if params == nil {
		params = map[string]string{}
	}
	token := opts.Token
	if token == "" {
		token = c.Token
	}
	jobID, err := uuidV4()
	if err != nil {
		return nil, &CloudiyError{"failed to generate job id: " + err.Error()}
	}

	reqBody, _ := json.Marshal(map[string]any{
		"job_id":         jobID,
		"kernel":         opts.Kernel,
		"input_data":     bytesToInts(opts.Data),
		"params":         params,
		"auth_token":     token,
		"consumer_pubkey": nil,
		"payment":        nilIfEmpty(opts.Payment),
	})

	req, err := http.NewRequest("POST", c.base+"/submit", bytes.NewReader(reqBody))
	if err != nil {
		return nil, &CloudiyError{err.Error()}
	}
	req.Header.Set("Content-Type", "application/json")
	if opts.Payment != "" {
		req.Header.Set("X-PAYMENT", opts.Payment)
	}

	res, err := c.http.Do(req)
	if err != nil {
		// Not auto-retried; surface the connection failure clearly.
		return nil, &CloudiyError{fmt.Sprintf("cannot reach node at %s (/submit): %v", c.base, err)}
	}
	body, _ := io.ReadAll(res.Body)
	res.Body.Close()

	if res.StatusCode == http.StatusPaymentRequired {
		var raw map[string]any
		if err := json.Unmarshal(body, &raw); err != nil {
			return nil, &CloudiyError{"402 with unparseable requirements: " + err.Error()}
		}
		return nil, newPaymentRequired(raw)
	}
	if res.StatusCode >= 300 {
		return nil, &CloudiyError{fmt.Sprintf("HTTP %d: %s", res.StatusCode, truncate(body, 300))}
	}

	var raw struct {
		JobID          string `json:"job_id"`
		OutputData     []int  `json:"output_data"`
		Status         string `json:"status"`
		ErrorMessage   string `json:"error_message"`
		ProviderPubkey string `json:"provider_pubkey"`
		Signature      string `json:"signature"`
		SignedBy       string `json:"signed_by"`
	}
	if err := json.Unmarshal(body, &raw); err != nil {
		return nil, &CloudiyError{"bad job response JSON: " + err.Error()}
	}
	if raw.Status == "error" {
		msg := raw.ErrorMessage
		if msg == "" {
			msg = "unknown error"
		}
		return nil, &CloudiyError{msg}
	}

	output := intsToBytes(raw.OutputData)

	var receipt map[string]any
	if h := res.Header.Get("x-payment-response"); h != "" {
		if dec, err := base64.StdEncoding.DecodeString(h); err == nil {
			_ = json.Unmarshal(dec, &receipt)
		}
	}

	// Verify the provider signature over (job_id, sha256(input), sha256(output))
	// — binds the output to THIS input. With a pin, the signer must also BE that
	// node; without it, a valid signature proves integrity but not intended node.
	verified := raw.Signature != "" && raw.SignedBy != "" &&
		(opts.ExpectPubkey == "" || raw.SignedBy == opts.ExpectPubkey) &&
		VerifyResult(raw.SignedBy, raw.JobID, opts.Data, output, raw.Signature)

	if !opts.NoVerify && !verified {
		switch {
		case raw.Signature == "" || raw.SignedBy == "":
			return nil, &SignatureError{"result was not signed by the provider — refusing to trust output (set NoVerify to accept unsigned)"}
		case opts.ExpectPubkey != "" && raw.SignedBy != opts.ExpectPubkey:
			return nil, &SignatureError{fmt.Sprintf("result signed by %s but expected %s — refusing to trust output", raw.SignedBy, opts.ExpectPubkey)}
		default:
			return nil, &SignatureError{"invalid provider signature — result may be tampered; refusing to trust output"}
		}
	}

	return &JobResult{
		JobID:             raw.JobID,
		Output:            output,
		Status:            raw.Status,
		ProviderPubkey:    raw.ProviderPubkey,
		PaymentReceipt:    receipt,
		SignatureVerified: verified,
		Signature:         raw.Signature,
		SignedBy:          raw.SignedBy,
	}, nil
}

// ToolSchema is an OpenAI/Anthropic-style function-tool definition.
type ToolSchema struct {
	Name        string         `json:"name"`
	Description string         `json:"description"`
	InputSchema map[string]any `json:"input_schema"`
}

// AsToolSchema returns a function-calling tool schema so LLM agents can invoke
// Cloudiy GPU compute as a tool. Pair with Client.Submit.
func AsToolSchema(node string) ToolSchema {
	if node == "" {
		node = "127.0.0.1:8080"
	}
	return ToolSchema{
		Name: "cloudiy_gpu_run",
		Description: "Run a compute kernel on a decentralized GPU (Cloudiy network, node " + node +
			"). Payment in USDC on Solana via x402. Kernels: vector_add ('a1,a2,...;b1,b2,...'), " +
			"matrix_mul ('m,k,n;A row-major;B row-major').",
		InputSchema: map[string]any{
			"type": "object",
			"properties": map[string]any{
				"kernel": map[string]any{"type": "string", "enum": []string{"vector_add", "matrix_mul"}},
				"data":   map[string]any{"type": "string", "description": "Kernel input in the documented format"},
			},
			"required": []string{"kernel", "data"},
		},
	}
}

// -- helpers ----------------------------------------------------------------

// uuidV4 returns a random RFC 4122 v4 UUID string using crypto/rand (no deps).
func uuidV4() (string, error) {
	var b [16]byte
	if _, err := rand.Read(b[:]); err != nil {
		return "", err
	}
	b[6] = (b[6] & 0x0f) | 0x40 // version 4
	b[8] = (b[8] & 0x3f) | 0x80 // variant 10
	return fmt.Sprintf("%x-%x-%x-%x-%x", b[0:4], b[4:6], b[6:8], b[8:10], b[10:16]), nil
}

// The node's JobRequest.input_data / JobResponse.output_data are JSON arrays of
// byte values (0-255), so bytes cross the wire as []int.
func bytesToInts(b []byte) []int {
	out := make([]int, len(b))
	for i, v := range b {
		out[i] = int(v)
	}
	return out
}

func intsToBytes(ns []int) []byte {
	out := make([]byte, len(ns))
	for i, n := range ns {
		out[i] = byte(n)
	}
	return out
}

func asFloat(v any) float64 {
	switch n := v.(type) {
	case float64:
		return n
	case json.Number:
		f, _ := n.Float64()
		return f
	case string:
		var f float64
		fmt.Sscanf(n, "%g", &f)
		return f
	default:
		return 0
	}
}

func asString(v any) string {
	s, _ := v.(string)
	return s
}

func nilIfEmpty(s string) any {
	if s == "" {
		return nil
	}
	return s
}

func truncate(b []byte, n int) string {
	if len(b) > n {
		return string(b[:n])
	}
	return string(b)
}
